bundle/jsky.app.ot/src/main/java/jsky/app/ot/gemini/editor/sitequality/TimingWindowDialog.java

```diff
@@ -0,0 +1,285 @@
+import java.awt.event.FocusEvent;
+import java.awt.event.KeyAdapter;
+import java.awt.event.KeyEvent;
+import java.text.*;
+import java.util.Date;
+import java.util.TimeZone;
+
+class TimingWindowDialog extends JDialog {
+    private static final long MS_PER_MINUTE = 1000 * 60;
+    private static final long MS_PER_HOUR   = MS_PER_MINUTE * 60;
+
+    private static final String INSTRUCTIONS =
+        "Specify a timing window. Starting time should be in the form YYYY-MM-DD hh:mm:ss " +
+        "and will be interpreted in UTC. Window duration and repeat period are specified in " +
+        "hours and minutes.";
+
+    private static final SimpleDateFormat UTC   = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss") {{
+        setTimeZone(TimeZone.getTimeZone("UTC"));
+    }};
+    private static final NumberFormat     TIMES = new DecimalFormat("######");
+
+
+    private final JFormattedTextField window;
+    private final JFormattedTextField duration;
+    private final JFormattedTextField period;
+    private final JFormattedTextField times;
+    private final ImList<JFormattedTextField> fields;
+
+    private final JRadioButton durationFixed;
+    private final JRadioButton durationForever;
+
+    private final JCheckBox repeatEnabled;
+
+    private final JRadioButton repeatTypeForever;
+    private final JRadioButton repeatTypeFixed;
+
+    private final ImList<JLabel> repeatLabels;
+
+    private final JButton okButton;
+    private final JButton cancelButton;
+
+    // Indicates if this dialog was cancelled.
+    private boolean cancelled;
+
+
+    TimingWindowDialog(final Frame owner) throws HeadlessException {
+        super(owner, "Edit Timing Window", true);
+        setResizable(false);
+        setDefaultCloseOperation(WindowConstants.DISPOSE_ON_CLOSE);
+        cancelled = true;
+
+        // Initialize components.
+        window = new JFormattedTextField(UTC);
+        window.setValue(new Date());
+
+        duration = new JFormattedTextField(HourMinuteFormat.HH_MM);
+        duration.setValue(MS_PER_HOUR * 24);
+
+        period = new JFormattedTextField(HourMinuteFormat.HH_MM_SS);
+        period.setColumns(8);
+        period.setValue(MS_PER_HOUR * 48);
+
+        times = new JFormattedTextField(TIMES);
+        times.setColumns(5);
+        times.setValue(1000);
+
+        fields = DefaultImList.create(window, duration, times, period);
+        fields.foreach(field -> {
+            field.setBorder(BorderFactory.createCompoundBorder(field.getBorder(), BorderFactory.createEmptyBorder(1, 3, 1, 3)));
+            field.addKeyListener(new KeyAdapter() {
+                @Override
+                public void keyReleased(KeyEvent e) {
+                    updateEnabledState();
+                }
+            });
+
+            // Whenever one of the fields loses focus, its contents are automatically set to the last valid edit,
+            // so the OK button should be enabled.
+            field.addFocusListener(new FocusAdapter() {
```

Comment:
This is new: the enabled status of the OK button relies on the contents of all `JFormattedTextFields` being acceptable, but the method to trigger an update on this only occurred when a `JFormattedTextField` received a `KeyListener.keyReleased` event.

This was not sufficient, as if a `JFormattedTextField` loses focus when it contains illegal contents, the contents are reset to their previous legal value; thus, to make the OK button enabled properly, we also need to enable it whenever a `JFormattedTextField` loses focus.